### PR TITLE
Maybe type classes

### PR DESCRIPTION
**Pull request by dwhitney**
Hi I enjoy using this library. It's really reduced a lot of boilerplate for me, so thanks so much for putting it together!

One thing that I wish it had out of the box is type class instances for `Maybe`. Using NullOrUndefined feels a bit leaky to me, and wrapping my `Maybe`s in `newtype`s just to get the functionality I wanted was becoming fairly tedious since I use `Maybe` so much. I also noticed there some similar sentiment on the latest PureScript Unscripted, so I decided to make this PR.

Things to note:
* when reading json, fields with `Maybe` types that have corresponding json values that are `null` or `undefined` will decode to a `Nothing`. Otherwise it will be `Just value`
* when writing `fieldName :: Just value` to json, the resulting json will be `"fieldName":"value"`, and when writing `fieldName :: Nothing`, the resulting json will be `"fieldName":null`
* I've added a foreign import to allow `null` to be written to the resulting json, and because I did that I didn't want it leaking out of the module, so I've added all of the relevant functions as exports.
* I've added four additional tests to test the edge cases and standard cases. 

Hope that you find this helpful. If you'd like any changes, I'm happy to make them!

### CODE REVIEW

File: src/Simple/JSON.purs [hunk 1]
```diff
@@ -73,6 +93,9 @@ instance readArray :: ReadForeign a => ReadForeign (Array a) where
 instance readNullOrUndefined :: ReadForeign a => ReadForeign (NullOrUndefined a) where
   readImpl = readNullOrUndefined readImpl
 
+instance readMaybe :: ReadForeign a => ReadForeign (Maybe a) where
+  readImpl foreignValue = (Foreign.readNullOrUndefined foreignValue) >>= (traverse readImpl)
```
Comment (jacereda) [inline src/Simple/JSON.purs]: What about this?

```
  readImpl = map unNullOrUndefined <<< readImpl
```

File: src/Simple/JSON.purs [hunk 2]
```diff
@@ -143,6 +166,12 @@ instance writeForeignArray :: WriteForeign a => WriteForeign (Array a) where
 instance writeForeignNullOrUndefined :: WriteForeign a => WriteForeign (NullOrUndefined a) where
   writeImpl (NullOrUndefined a) = maybe undefined writeImpl a
 
+instance writeForeignMaybe :: WriteForeign a => WriteForeign (Maybe a) where
+  writeImpl (Just a) = writeImpl a
+  writeImpl Nothing = null
```
Comment (jacereda) [inline src/Simple/JSON.purs]: I'd use `undefined` for `Nothing` to get smaller serialisation.

Comment (justinwoo) [reply src/Simple/JSON.purs]: I think a lot of existing implementations use Maybe -> null encoding though :cry: 

I think just trying to be compatible with Aeson and other libs would be good enough for Maybe, even though I would much prefer undefined too

File: src/Simple/JSON.purs [hunk 3]
```diff
@@ -143,6 +166,12 @@ instance writeForeignArray :: WriteForeign a => WriteForeign (Array a) where
...
+
+foreign import null :: Foreign
```
Comment (justinwoo) [inline src/Simple/JSON.purs]: Could we use the null from nullable? I'd like to not have to deal with FFI in this package...

File: src/Simple/JSON.purs (hunk 2 above)
Comment (jacereda) [reply src/Simple/JSON.purs]: Won't Aeson interpret a missing field as `Nothing`?

Comment (justinwoo) [reply src/Simple/JSON.purs]: Not afaict, it will just fail to decode if the field doesn't exist as `null`. But this is the encoding case, where writing `null` would match Aeson's `ToJSON` behavior, but yeah, I would prefer to write out undefined (missing fields in JSON) if this isn't a concern for @dwhitney

Comment (jacereda) [reply src/Simple/JSON.purs]: Looks like they use different functions depending on whether a missing key should be valid or not:

https://github.com/bos/aeson/blob/3a778548a928a5bb94a3defb526c985a4016e161/Data/Aeson/Types/FromJSON.hs#L689

File: src/Simple/JSON.purs (hunk 1 above)
Comment (dwhitney) [reply src/Simple/JSON.purs]: 👍

Comment (dwhitney) [reply src/Simple/JSON.purs]: will update the PR with this

File: src/Simple/JSON.purs (hunk 3 above)
Comment (dwhitney) [reply src/Simple/JSON.purs]: I thought about that and decided to go down this route in lieu of adding another dependency. That said, I'm fine with either.

File: src/Simple/JSON.purs (hunk 2 above)
Comment (dwhitney) [reply src/Simple/JSON.purs]: encoding `Nothing` to `null` is what I've encountered in the json serialization libs I've used. I also use GraphQL quite a bit and optional values are given the value of `null` when encoding json in those responses. I'd call it "industry standard" but don't quote me ;)

File: src/Simple/JSON.purs (hunk 3 above)
Comment (jacereda) [reply src/Simple/JSON.purs]: nullable is an indirect dependency anyway, better use that one.

Comment (justinwoo) [reply src/Simple/JSON.purs]: Ah, `null` isn't exposed by Nullable but really, I'd even prefer `toForeign $ toNullable Nothing`

**Comment by dwhitney** [comment]
done! travis should be doing it's work any moment

**Comment by jacereda** [comment]
Merged, thanks!
